core/src/main/java/net/opentsdb/query/hacluster/HAClusterFactory.java

```diff
@@ -173,417 +124,135 @@ public void setupGraph(final QueryPipelineContext context,
-          List<QueryNodeConfig> renamed_pushdowns =
-                  Lists.newArrayListWithExpectedSize(source_push_downs.size());
-          for (final QueryNodeConfig pd : source_push_downs) {
-            if (pd.getSources().contains(config.getId())) {
-              renamed_pushdowns.add(
-                      pd.toBuilder()
-                              .setSources(Lists.newArrayList(pushdown_id))
-                              .build());
-            } else {
-              renamed_pushdowns.add(pd);
-            }
-          }
-          new_sources.get(i).setPushDownNodes(renamed_pushdowns);
-          final Collection<String> sinks = pushDownSinks(renamed_pushdowns);
-          if (sinks.size() == 1) {
-            final String new_merger_id = sinks.iterator().next();
-            if (!merger.getId().equals(new_merger_id)) {
-              // only do this once or we get an exception.
-              MergerConfig new_merger = merger.toBuilder()
-                  .setDataSource(config.getId())
-                  .setId(sinks.iterator().next())
-                  .build();
-              planner.replace(merger, new_merger);
-              merger = new_merger;
-            }
-          }
-          final TimeSeriesDataSourceConfig new_source = (TimeSeriesDataSourceConfig) 
-              new_sources.get(i)
-              .build();
-          if (source_push_downs.size() == max_pushdowns) {
-            planner.addEdge(rebuilt, new_source);
-          } else {
-            final QueryNodeConfig mx = new_push_downs.get(
-                    max.get(source_push_downs.size()).getId());
-            planner.addEdge(mx, new_source);
-          }
-
-          if (context.query().isTraceEnabled()) {
-            context.queryContext().logTrace("Adding pushdown source: "
-                    + new_source.getSourceId());
-          }
-        }
-
-        // don't fall through!
-        return;
+      if (sources.get(i).factory().idType() != Const.TS_STRING_ID) {
+        needsIdConverter = true;
       }
+      newSources.add(localBuilder);
     }
 
-    // no push down, just replace
-    MergerConfig merger = MergerConfig.newBuilder()
-        .setAggregator(
-            Strings.isNullOrEmpty(builder.mergeAggregator()) ? 
-                tsdb.getConfig().getString(getConfigKey(AGGREGATOR_KEY)) : 
-                builder.mergeAggregator())
-        .addInterpolatorConfig(NumericInterpolatorConfig.newBuilder()
-          .setFillPolicy(FillPolicy.NONE)
-          .setRealFillPolicy(FillWithRealPolicy.NONE)
-          .setDataType(NumericType.TYPE.toString())
-          .build())
-        .setDataSource(config.getId())
-        .addSource(new_id)
-        .setId(config.getId())
-        .build();
+    MergerConfig merger = (MergerConfig) MergerConfig.newBuilder()
+            // TODO - want to make this configurable. Particularly when we're
+            // averaging or min-ing, etc.
+            .setAggregator("max")
```

Comment:
In case of getting data from multiple shards, this will have to be configurable (obtained from the group by) I guess.

Reply:
Yup